### PR TITLE
Support multiple undo/redo

### LINKED ISSUE #59: Implement complete undo and redo
- [ ] Final check on the undo and redo
- [ ] Amend the implementation to support more undoable commands if possible (annotate, etc)

Nice to have: 
- [x] Support undo/redo multiple commands (`undo 3`)
- [x] Currently, command history is not stored (only a list of states are stored in VersionedMark class). If possible, we can also store command history and display to users what command they are undoing/redoing.
- Support undo and redo even the application is closed (to implement this, history states of Mark must be stored somewhere and there must be a list of history states to be stored. (not going to do this because of memory)

### PR DESCRIPTION
**Pull request by Na-Nazhou**
Closes #59

### CODE REVIEW

File: src/main/java/seedu/mark/logic/commands/RedoCommand.java [hunk 1]
```diff
@@ -15,19 +15,39 @@
...
-    public static final String MESSAGE_FAILURE = "No more commands to redo!";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Redoes previous undo commands by the given number of steps. "
+            + "If not specified, the default number of steps is 1. \n"
+            + "Parameters: [STEP] (must be a positive integer)\n"
+            + "Example: " + COMMAND_WORD + " 2";
+
+    public static final String MESSAGE_SUCCESS = "Action(s) successfully redone: \n%1$s";
+    public static final String MESSAGE_FAILURE = "Not enough commands to redo! "
+            + "Only %1$s remaining undo command(s).\";";
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/commands/RedoCommand.java]:
```suggestion
            + "Only %1$s remaining undo command(s).";
```

File: src/main/java/seedu/mark/logic/commands/RedoCommand.java [hunk 2]
```diff
@@ -15,19 +15,39 @@
 
     public static final String COMMAND_WORD = "redo";
 
-    public static final String MESSAGE_SUCCESS = "Action successfully redone: \"%1$s\"";
-    public static final String MESSAGE_FAILURE = "No more commands to redo!";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Redoes previous undo commands by the given number of steps. "
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/commands/RedoCommand.java]:
```suggestion
            + ": Redoes a given number of previous undo commands. "
```

File: src/main/java/seedu/mark/logic/commands/RedoCommand.java [hunk 3]
```diff
@@ -15,19 +15,39 @@
...
+            + "If not specified, the default number of steps is 1. \n"
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/commands/RedoCommand.java]:
```suggestion
            + "If not specified, the default number of commands to redo is 1. \n"
```

File: src/main/java/seedu/mark/logic/commands/UndoCommand.java [hunk 4]
```diff
@@ -15,19 +15,39 @@
 
     public static final String COMMAND_WORD = "undo";
 
-    public static final String MESSAGE_SUCCESS = "Action successfully undone: \"%1$s\"";
-    public static final String MESSAGE_FAILURE = "No more commands to undo!";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Undoes previous undoable commands by the given number of steps. "
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/commands/UndoCommand.java]:
```suggestion
            + ": Undoes a given number of previous undoable commands. "
```

File: src/main/java/seedu/mark/logic/commands/UndoCommand.java [hunk 5]
```diff
@@ -15,19 +15,39 @@
...
+            + "If not specified, the default number of steps is 1. \n"
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/commands/UndoCommand.java]:
```suggestion
            + "If not specified, the default number of commands to undo is 1. \n"
```

File: src/main/java/seedu/mark/logic/parser/RedoCommandParser.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import seedu.mark.commons.util.StringUtil;
+import seedu.mark.logic.commands.RedoCommand;
+import seedu.mark.logic.parser.exceptions.ParseException;
+
+/**
+ * Parses input arguments and creates a new UndoCommand object
+ */
+public class RedoCommandParser implements Parser<RedoCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the GotoCommand
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/parser/RedoCommandParser.java]:
```suggestion
     * Parses the given {@code String} of arguments in the context of the RedoCommand
```

File: src/main/java/seedu/mark/logic/parser/RedoCommandParser.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+     * and returns a UndoCommand object for execution.
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/parser/RedoCommandParser.java]:
```suggestion
     * and returns a RedoCommand object for execution.
```

File: src/main/java/seedu/mark/logic/parser/UndoCommandParser.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import seedu.mark.commons.util.StringUtil;
+import seedu.mark.logic.commands.UndoCommand;
+import seedu.mark.logic.parser.exceptions.ParseException;
+
+/**
+ * Parses input arguments and creates a new UndoCommand object
+ */
+public class UndoCommandParser implements Parser<UndoCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the GotoCommand
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/parser/UndoCommandParser.java]:
```suggestion
     * Parses the given {@code String} of arguments in the context of the UndoCommand
```

File: src/main/java/seedu/mark/logic/parser/UndoCommandParser.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+     * @throws ParseException if the user input does not conform to the expected format
+     */
+    public UndoCommand parse(String args) throws ParseException {
+        try {
+            String trimmedArgs = args.trim();
+            if (trimmedArgs.isEmpty()) {
+                return new UndoCommand(1);
+            }
+            if (!StringUtil.isNonZeroUnsignedInteger(trimmedArgs)) {
+                throw new ParseException("Step is not a non-zero unsigned integer.");
```
Comment (dorcastan) [inline src/main/java/seedu/mark/logic/parser/UndoCommandParser.java]:
```suggestion
                throw new ParseException("The number of steps " + trimmedArgs + " is not a non-zero unsigned integer.");
```

File: src/main/java/seedu/mark/model/Model.java [hunk 10]
```diff
@@ -20,7 +20,9 @@
  * The API of the Model component.
  */
 public interface Model {
-    /** {@code Predicate} that always evaluate to true */
+    /**
+     * {@code Predicate} that always evaluate to true
```
Comment (dorcastan) [inline src/main/java/seedu/mark/model/Model.java]:
```suggestion
     * {@code Predicate} that always evaluates to true
```

File: src/main/java/seedu/mark/model/Model.java [hunk 11]
```diff
@@ -151,28 +158,41 @@
 
     /**
      * Returns true if the model has previous Mark states to restore.
+     *
+     * @param steps The number of steps to check
```
Comment (dorcastan) [inline src/main/java/seedu/mark/model/Model.java]:
```suggestion
     * @param steps The number of previous Mark states to check
```

File: src/main/java/seedu/mark/model/Model.java [hunk 12]
```diff
@@ -151,28 +158,41 @@
...
      */
-    boolean canUndoMark();
+    boolean canUndoMark(int steps);
+
+    int getMaxStepsToUndo();
 
     /**
      * Returns true if the model has undone Mark states to restore.
+     *
+     * @param steps The number of steps to check
```
Comment (dorcastan) [inline src/main/java/seedu/mark/model/Model.java]:
```suggestion
     * @param steps The number of undone Mark states to check
```

Comment (Na-Nazhou) [reply src/main/java/seedu/mark/model/Model.java]: Thanks!

File: src/main/java/seedu/mark/logic/commands/RedoCommand.java [hunk 13]
```diff
@@ -15,19 +15,39 @@
...
-    public static final String MESSAGE_FAILURE = "No more commands to redo!";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Redoes a given number of previous undo commands. "
+            + "If not specified, the default number of commands to redo is 1. \n"
+            + "Parameters: [STEP] (must be a positive integer)\n"
+            + "Example: " + COMMAND_WORD + " 2";
+
+    public static final String MESSAGE_SUCCESS = "Action(s) successfully redone: \n%1$s";
+    public static final String MESSAGE_FAILURE = "Not enough commands to redo! "
+            + "Only %1$s remaining undo command(s).";
```
Comment (openorclose) [inline src/main/java/seedu/mark/logic/commands/RedoCommand.java]: is this supposed to say redo?

maybe can say "only xxx remaining commands to undo"

File: src/main/java/seedu/mark/logic/commands/UndoCommand.java [hunk 14]
```diff
@@ -15,19 +15,39 @@
...
-    public static final String MESSAGE_FAILURE = "No more commands to undo!";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Undoes a given number of previous undoable commands. "
+            + "If not specified, the default number of commands to undo is 1. \n"
+            + "Parameters: [STEP] (must be a positive integer)\n"
+            + "Example: " + COMMAND_WORD + " 2";
+
+    public static final String MESSAGE_SUCCESS = "Action(s) successfully undone: \n%1$s";
+    public static final String MESSAGE_FAILURE = "Not enough commands to undo! "
+            + "Only %1$s remaining undoable command(s).";
```
Comment (openorclose) [inline src/main/java/seedu/mark/logic/commands/UndoCommand.java]: Yeah i think we should follow the same phrasing for redo

File: src/main/java/seedu/mark/logic/parser/RedoCommandParser.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
+package seedu.mark.logic.parser;
+
+import static seedu.mark.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+
+import seedu.mark.commons.util.StringUtil;
+import seedu.mark.logic.commands.RedoCommand;
+import seedu.mark.logic.parser.exceptions.ParseException;
+
+/**
+ * Parses input arguments and creates a new UndoCommand object
```
Comment (openorclose) [inline src/main/java/seedu/mark/logic/parser/RedoCommandParser.java]:
```suggestion
 * Parses input arguments and creates a new RedoCommand object
```

**Pull request merged**